### PR TITLE
Fixes for hue calculations for Color Control cluster

### LINKED ISSUE #8828: Hue value is calculated incorrectly when exceeds Max/Min value
When Device executing ColorControl Cluster commends `StepHue` and `MoveHue`, if the hue value reaches `(MAX_HUE_VALUE(0xFE) + 1)`, the result will be `0xFE + 1 - 0xFE = 1` (following current implementation of SDK).
But according to the Matter's Specification: `Increase the device’s hue by one step, in a continuous fashion. If the hue value reaches the maximum value then proceed to the minimum allowed value.`, that may mean `MAX_HUE_VALUE + 1 = MIN_HUE_VALUE`, so if the hue value reaches `(MAX_HUE_VALUE(0xFE) + 1)`, the result would be `0`.

There is also the same problem when subtracting hue value and exceeds MIN_HUE_VALUE .

How to reproduce:
```
chip-device-ctrl > zclread ColorControl CurrentHue 12344321 1 0
attributeValue: 249

chip-device-ctrl > zcl ColorControl StepHue 12344321 1 0 stepMode=1 stepSize=10 transitionTime=300 optionsMask=0 optionsOverride=0

zclread ColorControl CurrentHue 12344321 1 0
attributeValue: 5
```

### PR DESCRIPTION
**Pull request by mkardous-silabs**
#### Problem
* Calculation Error for finalHue when using EnhancedMoveHue with hue decreasing (moveMode = 3)
The incorrect finalHue is calculated when recalculting it when the first move hue is complete. This cause delta between two hues to be incorrect.
To reproduce:
`zcl ColorControl EnhancedMoveHue 12344321 1 0 moveMode=3 rate=100 optionsMask=0 optionsOverride=0`
* Fixes #8828 

#### Change overview
* For issue 8828, changes constants to max uint8 for wrap arround
* For finalhue error, switched initial and final hue in substract function

#### Testing
For finalHue error, 
Used following command to validate delta between two hues was correct after recalculation:
`zcl ColorControl EnhancedMoveHue 12344321 1 0 moveMode=3 rate=100 optionsMask=0 optionsOverride=0`

For issue 8828,
Used following commands to valiadate that wrap around was not skipping a value
`zcl ColorControl StepHue 12344321 1 0 stepMode=1 stepSize=10 transitionTime=300 optionsMask=0 optionsOverride=0`
`zcl ColorControl StepHue 12344321 1 0 stepMode=3 stepSize=10 transitionTime=300 optionsMask=0 optionsOverride=0`

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 1]
```diff
@@ -1611,7 +1611,7 @@ static uint8_t addHue(uint8_t hue1, uint8_t hue2)
 
     if (hue16 > MAX_HUE_VALUE)
     {
-        hue16 = static_cast<uint16_t>(hue16 - MAX_HUE_VALUE);
+        hue16 = static_cast<uint16_t>(hue16 - UINT8_MAX);
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1614]: Shouldn't this be `hue16 - MAX_HUE_VALUE - 1`, so we don't assume `MAX_HUE_VALUE == UINT8_MAX - 1`?

File: src/app/clusters/color-control-server/color-control-server.cpp [hunk 2]
```diff
@@ -1624,7 +1624,7 @@ static uint8_t subtractHue(uint8_t hue1, uint8_t hue2)
     hue16 = ((uint16_t) hue1);
     if (hue2 > hue1)
     {
-        hue16 = static_cast<uint16_t>(hue16 + MAX_HUE_VALUE);
+        hue16 = static_cast<uint16_t>(hue16 + UINT8_MAX);
```
Comment (bzbarsky-apple) [inline src/app/clusters/color-control-server/color-control-server.cpp:1627]: Similarly, this should be `hue16 + MAX_HUE_VALUE + 1`?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
This was a bug in the original Silicon Labs code, right?  @selissia @tecimovic could you please check whether there's a reason for the old (apparently buggy) behavior?

**Review by bzbarsky-apple — COMMENTED** [review]
Also, can we add automatic tests for this?

**Comment by mkardous-silabs** [comment]
> Also, can we add automatic tests for this?
Tests are already implemented in the yaml files for hue and enhanced hue variation. The only problem is, for them to actually check a variation, a wait/sleepFor needs to be added to the test suite. Right now, if we enable those tests, the test would start the hue viartion (moveHue, moveToHue, etc) and read the hue value right away before it has time to change.

**Comment by andy31415** [comment]
> > Also, can we add automatic tests for this?
> 
> Tests are already implemented in the yaml files for hue and enhanced hue variation. The only problem is, for them to actually check a variation, a wait/sleepFor needs to be added to the test suite. Right now, if we enable those tests, the test would start the hue viartion (moveHue, moveToHue, etc) and read the hue value right away before it has time to change.

Is there a TODO/issue for this? 
Would the test if adding wait/sleep have caught this bug? Is sleeping the only option or could we have 'only return success/failure after attribute has been applied'?

**Review by andy31415 — APPROVED** [review]

**Comment by mkardous-silabs** [comment]
> > > Also, can we add automatic tests for this?
> > 
> > 
> > Tests are already implemented in the yaml files for hue and enhanced hue variation. The only problem is, for them to actually check a variation, a wait/sleepFor needs to be added to the test suite. Right now, if we enable those tests, the test would start the hue viartion (moveHue, moveToHue, etc) and read the hue value right away before it has time to change.
> 
> Is there a TODO/issue for this?

I don't know if there is an issue/TODO for adding wait/sleeps in the test suite. All I know is that there have been conversations about adding them.

> Would the test if adding wait/sleep have caught this bug? 

A test case checking the rollover at MAX_HUE and MIN_HUE for the max/min hue calculations issue and another test case that verifies that steps between hue updates remains the same after the recalculation should have detected the bugs. These tests would be complicated to implement since we need to wait a certain amount of time before checking the `currentHue` and the wait/sleep won't necessarly be exact in the timeout time.

> Is sleeping the only option or could we have 'only return success/failure after attribute has been applied'?

To evaluate that the hue variations are correct, we need a sleep/wait to be able to the let the functions (moveHue, moveTohue, etc) iterate a few times. A success/failure after attribute application is only usefull to check if attributs were properly written and not if all the hue variation calculations are correct.

**Comment by bzbarsky-apple** [comment]
>  The only problem is, for them to actually check a variation, a wait/sleepFor needs to be added to the test suite

Looking at the spec StepHue can send a TransitionTime of 0, which I would expect to take effect immediately... but our implementation uses 1 if 0 is sent.  Is the implementation wrong here, or the spec, or is something something else going on?

@Rob-Houtepen-Signify @selissia

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by Rob-Houtepen-Signify** [comment]
> > The only problem is, for them to actually check a variation, a wait/sleepFor needs to be added to the test suite
> 
> Looking at the spec StepHue can send a TransitionTime of 0, which I would expect to take effect immediately... but our implementation uses 1 if 0 is sent. Is the implementation wrong here, or the spec, or is something something else going on?
> 
Too me sounds like an implementation detail, where with 1 some code is executed and stopped at 0. I would not mind too much.

**Comment by bzbarsky-apple** [comment]
> Too me sounds like an implementation detail

It's a spec-observable difference, right?  If I do a StepHue with a TransitionTime of 0, does the spec allow a subsequent read of hue to get the old value?  It needs to clearly define the behavior....

**Comment by jmartinez-silabs** [comment]
@bzbarsky-apple @andy31415 @Rob-Houtepen-Signify @mkardous-silabs Even if we do transitiontime 0, the test could validate the final result (final value of the hue or enhanced Hue) but it would not validated the transition or calculation of the steps/rates. It becomes essentially a write attributes to hue in the received command callback.

Right now when we receive such commands we set our transition states and calculate de rate/steps to be done during that transition time and kick of the state machine. The state machine then establish the transition every 100ms.

Not sure what is the best way to test those transition and corner case in the unit test

**Pull request merged**